### PR TITLE
Rados early automation of EC 8+6 pools with MSR rules

### PR DESCRIPTION
**Pull request by pdhiran**
1. Added initial test cases for EC 8+6 pool with MSR rules
2. Added suite and conf file for 3 AZ cluster deployment

### CODE REVIEW

File: suites/squid/rados/tier-3_rados_test-4-node-8+6-msr-ecpools.yaml [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        plugin: jerasure
+        create_rule: true
+        pre_create_rule: false
+        crush-failure-domain: host
+        change_subtree_limit: true
+        crush-osds-per-failure-domain: 4
+        crush-num-failure-domains: 4
+        delete_pools:
+          - test_ec_pool
+      desc: 2+2@4 EC pool life cycle with serviceability scenarios
```
Comment (neha-gangadhar) [inline suites/squid/rados/tier-3_rados_test-4-node-8+6-msr-ecpools.yaml:168]: desc needs to be modified to 8+6

**Comment by neha-gangadhar** [comment]
Can we add all the tests we currently have for both EC 2+2 and 8+6 -> Lets discuss offline and add existing automated scenarios.

File: suites/squid/rados/tier-3_rados_test-4-node-8+6-msr-ecpools.yaml (hunk 1 above)
Comment (pdhiran) [reply suites/squid/rados/tier-3_rados_test-4-node-8+6-msr-ecpools.yaml]: Updated.

**Comment by pdhiran** [comment]
@neha-gangadhar , all the requested changes are made. Please review.

File: ceph/rados/pool_workflows.py [hunk 2]
```diff
@@ -683,3 +683,221 @@ def fetch_pool_stats(self, pool: str) -> dict:
...
+                    "Checking the ideal PG count on the pool and the threshold set. default is 3"
+                )
+                pool_data = self.rados_obj.get_pg_autoscale_status(pool_name=pool)
+                final_pg_num = pool_data["pg_num_final"]
+                log.debug(
+                    f"Pool : {pool} details as fetched from autoscale-status : {pool_data}"
+                )
+                # for the threshold var, refer :
+                # https://docs.ceph.com/en/latest/rados/operations/placement-groups/#viewing-pg-scaling-recommendations
+                if float(final_pg_num / init_pg_count) <= 3.0:
```
Comment (SrinivasaBharath) [inline ceph/rados/pool_workflows.py:772]: Could you please explain the logic about the threshold value?  
I checked another document to understand the details.
https://ceph.io/en/news/blog/2022/autoscaler_tuning/

File: ceph/rados/pool_workflows.py [hunk 3]
```diff
@@ -683,3 +683,221 @@ def fetch_pool_stats(self, pool: str) -> dict:
...
+                    log.info(
+                        "PG count not expected to increase as the final pg num is less than the threshold"
+                    )
+                    if modify_threshold:
+                        log.info(
+                            "Param passed to modify the threshold set. Changing it to 1.3 from 3."
+                        )
+                        cmd = "ceph osd pool set threshold 1.3"
```
Comment (SrinivasaBharath) [inline ceph/rados/pool_workflows.py:780]: The same as above.

File: ceph/rados/pool_workflows.py [hunk 4]
```diff
@@ -683,3 +683,221 @@ def fetch_pool_stats(self, pool: str) -> dict:
...
+                    )
+                    return False, inactive_pg
+                log.info(
+                    "PGs increased to desired levels after application of bulk flag"
+                )
+                if modify_threshold:
+                    log.info(
+                        "Param passed to modify the threshold set. Reverting it back to 3"
+                    )
+                    cmd = "ceph osd pool set threshold 3"
```
Comment (SrinivasaBharath) [inline ceph/rados/pool_workflows.py:830]: Please create a method in the common module.

**Review by SrinivasaBharath — CHANGES_REQUESTED** [review]

File: conf/squid/rados/test-confs/pool-configurations.yaml [hunk 5]
```diff
@@ -27,9 +27,38 @@ erasure:
...
     crush-failure-domain: osd
   sample-pool-3:
-    pool_name: ec_pool_30
+    profile_name: ec86_profile
+    pool_name: ec86_msr_pool
     pool_type: erasure
     k: 8
     m: 4
+    create_rule: true
+    crush-osds-per-failure-domain: 4
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply conf/squid/rados/test-confs/pool-configurations.yaml:36]: There already exists pools with that scenario in the same config file. please review.

File: ceph/rados/core_workflows.py [hunk 6]
```diff
@@ -1100,46 +1100,102 @@ def create_erasure_pool(self, name: str, **kwargs) -> bool:
...
+        create_rule = kwargs.get("create_rule", False)
         plugin = kwargs.get("plugin", "jerasure")
         pool_name = kwargs.get("pool_name")
         force = kwargs.get("force")
-        profile_name = f"ecprofile_{name}"
+        profile_name = kwargs.get("profile_name", f"ecp_{name}")
+        rule_name = f"rule_{name}"
 
         # Creating an erasure coded profile with the options provided
         cmd = (
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/core_workflows.py:1126]: new method for what? this is already method for creating EC pools.

File: ceph/rados/core_workflows.py [hunk 7]
```diff
@@ -1100,46 +1100,102 @@ def create_erasure_pool(self, name: str, **kwargs) -> bool:
...
             cmd = cmd + f" d={d}"
         if force:
             cmd = cmd + " --force"
+
+        log.debug(f"Final command to create EC pool : {cmd}")
         try:
-            self.node.shell([cmd])
+            self.run_ceph_command(cmd=cmd)
+            time.sleep(5)
+            profile_ls = "ceph osd erasure-code-profile ls"
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/core_workflows.py:1154]: added new method

File: ceph/rados/core_workflows.py [hunk 8]
```diff
@@ -1100,46 +1100,102 @@ def create_erasure_pool(self, name: str, **kwargs) -> bool:
...
+
+        # Creating the Crush rule for the profile created
+        if create_rule:
+            cmd = f"ceph osd crush rule create-erasure {rule_name} {profile_name}"
+            self.run_ceph_command(cmd=cmd)
+            time.sleep(5)
+
+            # Creating the pool with the profile created
+            # https://docs.ceph.com/en/latest/rados/operations/crush-map/#rules
+            cmd = "ceph osd crush rule ls"
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/core_workflows.py:1177]: added new method

File: ceph/rados/core_workflows.py [hunk 9]
```diff
@@ -1100,46 +1100,102 @@ def create_erasure_pool(self, name: str, **kwargs) -> bool:
...
-        if not self.create_pool(
-            ec_profile_name=profile_name,
-            **kwargs,
-        ):
-            log.error(f"Failed to create Pool {pool_name}")
-            return False
+
+        # Creating the Crush rule for the profile created
+        if create_rule:
+            cmd = f"ceph osd crush rule create-erasure {rule_name} {profile_name}"
```
[the remark this answers is not in the document]
Comment (pdhiran) [reply ceph/rados/core_workflows.py:1171]: This should be part of the create_erasure pool module, cannot be a new module. 
This is very specific to EC pools, and only created if EC profile needs a explicit rule to be created.

Not creating a new method for this. Let's meet for discussion if needed.

File: ceph/rados/pool_workflows.py (hunk 2 above)
Comment (pdhiran) [reply ceph/rados/pool_workflows.py:772]: What are the queries that you have with regard to Threshold then? The explanation given in both the doc atatched as comment and the doc that you have attached provide sufficient explanation as to what does the threshold value does.

File: ceph/rados/pool_workflows.py (hunk 3 above)
Comment (pdhiran) [reply ceph/rados/pool_workflows.py:780]: what is the query ?

File: ceph/rados/pool_workflows.py (hunk 4 above)
Comment (pdhiran) [reply ceph/rados/pool_workflows.py:830]: New method added.

**Comment by pdhiran** [comment]
@SrinivasaBharath , Requested changes made. Please review.

File: ceph/rados/pool_workflows.py [hunk 10]
```diff
@@ -683,3 +683,230 @@ def fetch_pool_stats(self, pool: str) -> dict:
...
+                    log.info(
+                        "Sleeping for 60 secs and checking the PG states and PG count"
+                    )
+                    time.sleep(60)
+                else:
+                    log.error(
+                        f"pg_num on pool {pool} did not reach the desired levels of PG count "
+                        f"with bulk flag enabled"
+                        f"Expected : {pg_count_bulk_true}"
+                    )
```
Comment (SrinivasaBharath) [inline ceph/rados/pool_workflows.py:831]: Please check that the function returns without resetting the default threshold value which was assigned 1.3 in the previous steps.

**Comment by SrinivasaBharath** [comment]
The build is not available to check the execution logs. The logic and functionality look good to me.

File: ceph/rados/pool_workflows.py (hunk 10 above)
Comment (pdhiran) [reply ceph/rados/pool_workflows.py:831]: fixed it. reverting back in finally in case of failures.

**Review by neha-gangadhar — APPROVED** [review]
Approved

**Review by SrinivasaBharath — APPROVED** [review]
LGTM

**Pull request merged**